peer-ids/peer-ids.md

```diff
@@ -91,154 +101,160 @@ files.
+then use [standard Bitcoin
+encoding](https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#der-encoding).
 
-In all cases, implementation MAY allow the user to enable/disable specific key types via configuration.
-Note that disabling support for compulsory key types may hinder connectivity.
+#### ECDSA
 
-Keys are encoded into byte arrays and serialized into the `Data` field of the
-protobuf messages described above.
+We encode the public key using ASN.1 DER.
 
-The following sections describe each key type's encoding rules.
+We encode the private key using DER-encoded PKIX.
 
-libp2p implementations MUST use the encoding described below when embedding
-keys into the `PublicKey` and `PrivateKey` messages described above, whether
-for transmission, Peer Id generation, or storage.
+To sign a message, we hash the message with SHA 256, and then sign it with the
+[ECDSA standard algorithm](https://tools.ietf.org/html/rfc6979), then we encode
+it using [DER-encoded ASN.1.](https://wiki.openssl.org/index.php/DER)
 
-Implementations can use whatever in-memory representation is convenient,
-provided the encodings described below are used at the "I/O boundary".
+## Peer Ids
 
-In addition to key encodings, the sections below cover the signing method used
-when signing and verifying messages with each key type.
+Peer IDs are derived by hashing the encoded public key with
+[multihash][multihash]. Keys that serialize to more than 42 bytes must be hashed
+using sha256 multihash, keys that serialize to at most 42 bytes must be hashed
+using the "identity" mulish codec.
 
-### RSA
+Specifically, to compute a peer ID of a key:
 
-We encode the public key using the DER-encoded PKIX format.
+1. Encode the public key as described in the [keys](#keys) section.
+4. If the length of the serialized bytes is less than or equal to 42, compute
+   the "identity" multihash of the serialized bytes. In other words, no
+   hashing is performed, but the [multihash format is still
+   followed][multihash] (byte plus varint plus serialized bytes). The idea
+   here is that if the serialized byte array is short enough, we can fit it in
+   a multihash verbatim without having to condense it using a hash function.
+5. If the length is greater than 42, then hash it using it using the SHA256
+   multihash.
 
-We encode the private key as a PKCS1 key using ASN.1 DER.
+### String representation
 
-To sign a message, we first hash it with SHA-256 and then sign it using the [RSASSA-PKCS1-V1.5-SIGN](https://tools.ietf.org/html/rfc3447#section-8.2) method, as originally defined in [RSA PKCS#1 v1.5](https://tools.ietf.org/html/rfc2313).
+There are two ways to represent peer IDs in text: as a raw
+[base58btc][base58btc] encoded multihash (e.g., `Qm...`, `1...`) and as a
+[multibase][multibase] encoded [CID][cid] (e.g., `bafz...`). Libp2p is slowly
+transitioning from the first (legacy) format to the second (new).
```

Comment:
The first one is CIDv0, the second one is CIDv1+, it may help to specify it as that.

Reply:
It's not, that's one of the reasons I modified the spec. The first one is a bare multihash, the second one is a CID.

The confusing part is that CIDv0s are _also_ bare multihashes. However, a CIDv0 is defined as a base58 encoded multihash of a dag-pb protobuf. A libp2p key is not a dag-pb protobuf so a "v0" peer ID is _not_ a CIDv0 CID, it just happens to look like one.